### PR TITLE
Laboratory work #2, Egorova Anastasia - 19FPL2

### PR DESCRIPTION
**Pull request by egorova-ai19-2**

### CODE REVIEW

File: lab_2/main.py [hunk 1]
```diff
@@ -34,7 +55,36 @@ def fill_lcs_matrix(first_sentence_tokens: tuple, second_sentence_tokens: tuple)
...
+            if first_sentence_tokens[first_sent_word - 1] == second_sentence_tokens[second_sent_word - 1]:
+                lcs_matrix[first_sent_word][second_sent_word] = lcs_matrix[first_sent_word - 1][second_sent_word - 1] \
+                                                                + 1
+            else:
+                lcs_matrix[first_sent_word][second_sent_word] = max(lcs_matrix[first_sent_word - 1][second_sent_word],
+                                                                    lcs_matrix[first_sent_word][second_sent_word - 1])
+
+    del lcs_matrix[0]
+    for row in lcs_matrix:
+        del row[0]
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Please, leave some explanation.

**Review by WhiteJaeger — CHANGES_REQUESTED** [review]
Great job so far!

File: lab_2/main.py (hunk 1 above)
Comment (egorova-ai19-2) [reply lab_2/main.py]: In this function zero matrix is created with one row and one column more than lengths of sentences: we need something to substract from when filling the matrix for first and last element of each row (as there is no number). After this, zero row and zero column are being removed, so filled matrix corresponds lengths of sentences.

Comment (WhiteJaeger) [reply lab_2/main.py]: Could not it be done without adding empty column and row?

File: lab_2/main.py [hunk 2]
```diff
@@ -92,7 +229,33 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+
+    if is_type_incorrect or not all(original_sentence_tokens) or not all(suspicious_sentence_tokens) or not all(lcs):
+        return ()
+
+    indexes_diff_original = [original_sentence_tokens.index(word) for word in original_sentence_tokens
+                             if word not in lcs]
+    indexes_diff_suspicious = [suspicious_sentence_tokens.index(word) for word in suspicious_sentence_tokens
+                               if word not in lcs]
+
+    def find_diff_indexes(indexes_diff_words):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: Place this function outside and use here as an utility one.

File: lab_2/main.py [hunk 3]
```diff
@@ -92,7 +229,33 @@ def find_diff_in_sentence(original_sentence_tokens: tuple, suspicious_sentence_t
...
+        changes_indexes = []
+        for index, diff_word_index in enumerate(indexes_diff_words):
+            changes_indexes.append(diff_word_index)
+            if (index != 0) and (diff_word_index - indexes_diff_words[index - 1] == 1):
```
Comment (WhiteJaeger) [inline lab_2/main.py]: `index != 0` can be shortened.

**Review by WhiteJaeger — COMMENTED** [review]
Good job!

File: lab_2/main.py (hunk 1 above)
Comment (egorova-ai19-2) [reply lab_2/main.py]: I can think only of if/elif/else and checking if row or column are first (with index = 0)

File: lab_2/main.py [hunk 4]
```diff
@@ -24,7 +51,14 @@ def create_zero_matrix(rows: int, columns: int) -> list:
     e.g. rows = 2, columns = 2
     --> [[0, 0], [0, 0]]
     """
-    pass
+    is_rows_columns_correct = ((not isinstance(rows, bool)) and (not isinstance(columns, bool)) and
+                               isinstance(rows, int) and isinstance(columns, int) and (rows > 0) and (columns > 0))
+    if not is_rows_columns_correct:
+        return []
+
+    zero_matrix = [[0] * columns for i in range(rows)]
```
Comment (WhiteJaeger) [inline lab_2/main.py]: There is a convention to use `_` if value will not be used.

`[[0] * columns for _ in range(rows)]`

File: lab_2/main.py [hunk 5]
```diff
@@ -57,7 +138,27 @@ def find_lcs(first_sentence_tokens: tuple, second_sentence_tokens: tuple, lcs_ma
...
+
+    lcs = []
+    for index_r, row in enumerate(reversed(lcs_matrix)):
+        for index_col, column in enumerate(reversed(row)):
+            if not row or not column:
+                return ()
+            if first_sentence_tokens[index_r] == second_sentence_tokens[index_col]:
+                lcs.append(second_sentence_tokens[index_col])
+
+    lcs = tuple(lcs)
```
Comment (WhiteJaeger) [inline lab_2/main.py]: You can place it directly in the `return`.

**Review by WhiteJaeger — COMMENTED** [review]
Good work.
Just some minor fixes and you are done here.

**Review by WhiteJaeger — APPROVED** [review]
Great!
You are done here.

**Pull request closed without merging**